### PR TITLE
[#issue 222]cache module

### PR DESCRIPTION
**Pull request by vidhya567**
-A separate angular cache module is created
-In mapcityJson function just calls the getCityPolygonsFromCache function of cache,every other function of Cache is abstracted away from map
Cache replacement
-divide the cachedRegion into parts excluding the overlapping part of current Polygon region request
-cut the bigger parts first see whether deleteTarget is achieved
-if not cut the next part

### CODE REVIEW

File: twittermap/conf/application.conf [hunk 1]
```diff
@@ -78,5 +78,5 @@ bounded-mailbox {
 # for the debug purpose, we can load a smaller json to speed up the front-end
 cloudberry.register = "http://localhost:9000/admin/register"
 cloudberry.ws = "ws://localhost:9000/ws"
-us.city.path = "/public/data/city.sample.json"
+us.city.path = "/public/data/city.json"
```
Comment (waans11) [inline twittermap/conf/application.conf]: @JavierJia mentioned that this needs to be reverted. @JavierJia: can you confirm?

File: twittermap/public/javascripts/cache/services.js [hunk 2]
```diff
@@ -0,0 +1,322 @@
+ angular.module('cloudberry.cache', ['leaflet-directive', 'cloudberry.common' ])
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Let's put a simple comment about this cache module - what it does?

File: twittermap/public/javascripts/cache/services.js [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+  var cacheSize = 0;
+  var insertedTreeIDs = new Set();
+  var cacheThreshold = 7500;//hard limit
+  var DeleteTarget = 0;
+  var DeletedCount = 0;
+  var preFetchDistance = 25;
+
+
+  /*Call will happen from map controller and this functions sees whether a required data is present in cache or not
+  if not gets from midleware*/
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Let's use the following:

Map controller calls this function and this function checks whether a requested region is present in the cache or not. If not, it gets the requested region data from the middleware.

File: twittermap/public/javascripts/cache/services.js [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+       var currentRequestPolygon = turf.polygon([[
+                                    [bounds._northEast.lng,bounds._northEast.lat],
+                                    [bounds._northEast.lng,bounds._southWest.lat],
+                                    [bounds._southWest.lng,bounds._southWest.lat],
+                                    [bounds._southWest.lng,bounds._northEast.lat],
+                                    [bounds._northEast.lng,bounds._northEast.lat]
+                                ]]);
+
+       var bbox = turf.bbox(currentRequestPolygon);
+        // to search in Rbush Tree we need this in this format
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: // to search in Rbush tree, we need the MBR of the requested region.

File: twittermap/public/javascripts/cache/services.js [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+       var item = {
+           minX : bbox[0],
+           minY : bbox[1],
+           maxX : bbox[2],
+           maxY : bbox[3]
+       }
+
+
+       if(typeof cachedRegion != "undefined" && typeof turf.difference(currentRequestPolygon,cachedRegion) == "undefined") {
+              //cache HIT
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: I think @JavierJia also mentioned about this. @vidhya567 : can you check his comments on the previous commit and apply them?

File: twittermap/public/javascripts/cache/services.js [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+
+             var result = cachedCityPolygonTree.search(item);
+             data_response = turf.featureCollection(result);
+             deferred.resolve(data_response);
+
+             return deferred.promise();
+
+       }else{
+                   //cache MISS
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Un-indent one more level to align with the following statements.

File: twittermap/public/javascripts/cache/services.js [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+
+                      });
+              }).error(function(data) {
+                                    console.error("Load city data failure");
+              });
+              return deferred.promise();
+       }
+  }
+//to insert polygons into the Rtree
+ var insertIntoTree = function insertIntoTree(features,currentRequest){
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Can you add a comment about features? What is it?

File: twittermap/public/javascripts/cache/services.js [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+       }
+
+       if(cacheSize >= cacheThreshold){
+
+             DeleteTarget = cacheSize - cacheThreshold;
+             evict(currentRequest).done(function(){
+
+                   cachedCityPolygonTree.load(nodes);
+                   cacheSize = nodes.length;
+                   console.log(" Size:",cacheSize);
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Remove this log.

File: twittermap/public/javascripts/cache/services.js [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+              });
+          deferred.resolve();
+          return deferred.promise();
+
+       }else{
+               cachedCityPolygonTree.load(nodes);
+               console.log(" Size:",cacheSize);
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Remove this log.

File: twittermap/public/javascripts/cache/services.js [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+            })
+            return deferred.promise();
+
+       }else if(!LowerRight &&   !LowerLeft && !UpperLeft && !UpperRight){
+
+              cutRegion(C_minX,C_minY,C_maxX,C_maxY,true,true).done(function(){
+              //NO Overlap
+                     deferred.resolve();
+              }).fail(function(){
+//                      cacheThreshold += DeleteTarget;
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Delete this commented statement.

File: twittermap/conf/application.conf (hunk 1 above)
Comment (vidhya567) [inline twittermap/conf/application.conf]: Oh I was testing thats why,I will change

File: twittermap/public/javascripts/cache/services.js (hunk 5 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Yes I have changed them from last time.I added typeof.This is the most safest way in javascript to see if a variable is not initialized with any value.I saw 4 or 5 stack overflow posts to make sure I write the correct syntax

File: twittermap/public/javascripts/cache/services.js (hunk 6 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Changed Indents,Is it fine?

File: twittermap/public/javascripts/cache/services.js [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+                     deferred.resolve();
+              }).fail(function(){
+
+                      clearCache().done();
+                      deferred.resolve();
+              })
+              return deferred.promise();
+       }
+ }
+//sees whether cutting the region can satisfy the target ,If not sends a failed message to evict function
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Please rephrase the comment: "Checks whether evicting some city polygons in a part of the cache region satisfies the target deletion count. If not, returns a failure message to the caller."

File: twittermap/public/javascripts/cache/services.js [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+//X = true Y = false
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Can you rephrase the comment like this (other comments too)? XorY: Whether evicting the cached region vertically or horizontally. true - X axis (horizontally), false - Y axis (vertically)

File: twittermap/public/javascripts/cache/services.js [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+//TopTo Bottom = true
+//BottomToTop = false
+//LeftToRight = true
+//RightToLeft = false
+ var cutRegion =  function findCornerofEviction(minX,minY,maxX,maxY,XorY,Direction){
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: XorY -> isHorizontalEviction

File: twittermap/public/javascripts/cache/services.js [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+                if(Direction){
+                    //Left to Right
+                    line = turf.lineString([[minX,maxY],[maxX,maxY]]);
+                }else{
+                    //Right to Left
+                    line = turf.lineString([[maxX,maxY],[minX,maxY]]);
+                }
+            }else{
+                if(Direction){
+                    //TopToBottom
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Top to bottom

File: twittermap/public/javascripts/cache/services.js [hunk 15]
```diff
@@ -0,0 +1,321 @@
...
+                    line = turf.lineString([[maxX,maxY],[maxX,minY]]);
+                }else{
+                    //BottomToTop
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Bottom to top

File: twittermap/public/javascripts/cache/services.js [hunk 16]
```diff
@@ -0,0 +1,321 @@
...
+                    line = turf.lineString([[maxX,minY],[maxX,maxY]]);
+                }
+            }
+
+            var distance = turf.lineDistance(line, 'miles');
+            var Move = distance/10;
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Makes 10 slices.

File: twittermap/public/javascripts/cache/services.js [hunk 17]
```diff
@@ -0,0 +1,321 @@
...
+            var start = 0;
+            var stop= distance;
+            var removeItems;
+            var sliced,cutPoint,cutBbox,remove_search;
+
+            while(DeletedCount<DeleteTarget){
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: DeletedCount --> deletedPolygonCount

File: twittermap/public/javascripts/cache/services.js [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+                                        minX: cutBbox[0],
+                                        minY: cutBbox[1],
+                                        maxX: cutBbox[2],
+                                        maxY: cutBbox[3]
+                                    }
+                  removeItems = cachedCityPolygonTree.search(remove_search);
+                  DeletedCount = removeItems.length;
+                  Move += Move;
+                  if(Move>stop)
+                  {break;}
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Doesn't this calculate the calculated region again and again? For example, if there are 10 slices (1,2,3,4,5,6,7,8,9,10) and suppose deleting 5 regions (up to 5) satisfies the condition.  then, it looks like the first region 1 is searched 5 times. Is this right? Why can't we do an incremental calculation?

File: twittermap/public/javascripts/cache/services.js [hunk 19]
```diff
@@ -0,0 +1,321 @@
...
+                  deferred.reject();
+                  return deferred.promise();
+            }else{
+                  // SucessFul Delete
+                  deferred.resolve();
+                  return deferred.promise();
+            }
+ }
+
+//wher deletion from tree occurs
```
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: Rephrase this comment.

File: twittermap/public/javascripts/cache/services.js [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
+
+                                                var result = cachedCityPolygonTree.search(item);
+                                                data_response = turf.featureCollection(result);
+
+                                                deferred.resolve(data_response);
+
+
+                      });
+              }).error(function(data) {
+                                    console.error("Load city data failure");
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: This was something that was present in earlier codebase whenever They were sending HTTP requests ,They have this error function if middleware had failed.

Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Could you see the map/controller.js file ,function loadGeoJsonFiles,There is a similar HTTP get request in this function.

File: twittermap/public/javascripts/cache/services.js [hunk 21]
```diff
@@ -0,0 +1,321 @@
...
+       var treeID;
+
+       for(var id in features){
+
+           var box = turf.bbox(features[id]);
+           features[id].minX = box[0];
+           features[id].minY = box[1];
+           features[id].maxX = box[2];
+           features[id].maxY = box[3];
+           features[id].properties["centerLog"] = (features[id].maxX + features[id].minX) / 2;
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: It is a field we have in format of our geo JSONs .Could you see the loadCityJsonByBound I have removed setCenterAndBoundary and written a comment there,This function does the same work as setCenterAndBoundary does along with what is needed for Cache.We talked about this some time ago.

File: twittermap/public/javascripts/cache/services.js (hunk 18 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: You are right ,But  We have libraries to search incrementally,it is implementation wise very hard and time consuming to add regions(merge is expensive),Thats why search for bigger region each time .This sounded much better when time needed to be low.

File: twittermap/public/javascripts/cache/services.js [hunk 22]
```diff
@@ -0,0 +1,321 @@
...
+
+       }else{
+               cachedCityPolygonTree.load(nodes);
+               deferred.resolve();
+               return deferred.promise();
+       }
+
+ }
+
+//X = true Y = false
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: removed

File: twittermap/public/javascripts/cache/services.js [hunk 23]
```diff
@@ -0,0 +1,321 @@
...
+//TopTo Bottom = true
+//BottomToTop = false
+//LeftToRight = True
+//RightToLeft = False
+//determine which region of cached region to cut to satisfy Target,if first region couldn't satisfy the Target go to new region
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Changed

File: twittermap/public/javascripts/cache/services.js [hunk 24]
```diff
@@ -0,0 +1,321 @@
...
+       var C_maxY = cache_bbox[3];
+
+       var request_bbox = turf.bbox(currentRequest);
+       var R_minX = request_bbox[0];
+       var R_minY = request_bbox[1];
+       var R_maxX = request_bbox[2];
+       var R_maxY = request_bbox[3];
+
+       var CacheMBR = turf.bboxPolygon(cache_bbox);
+       //Four Corners of Request to see which is inside cache_region
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Revised

File: twittermap/public/javascripts/cache/services.js (hunk 18 above)
Comment (waans11) [inline twittermap/public/javascripts/cache/services.js]: I wonder whether you have any experimental results for this.

File: twittermap/public/javascripts/cache/services.js [hunk 25]
```diff
@@ -0,0 +1,321 @@
...
+and provide it to the user without sending http request.*/
+angular.module('cloudberry.cache', ['leaflet-directive', 'cloudberry.common' ])
+ .service('Cache', function( $window, $http, $compile){
+
+  var cachedCityPolygonTree = rbush();
+  var cachedRegion ;
+  var cacheSize = 0;
+  var insertedTreeIDs = new Set();
+  var cacheThreshold = 7500;//hard limit
+  var DeleteTarget = 0;
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Changed

File: twittermap/public/javascripts/cache/services.js [hunk 26]
```diff
@@ -0,0 +1,321 @@
...
+  var DeletedCount = 0;
+  var preFetchDistance = 25;
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Added comment

File: twittermap/public/javascripts/cache/services.js [hunk 27]
```diff
@@ -0,0 +1,321 @@
...
+             var result = cachedCityPolygonTree.search(item);
+             data_response = turf.featureCollection(result);
+             deferred.resolve(data_response);
+
+             return deferred.promise();
+
+       }else{
+                   //cache MISS
+
+              var buffered = turf.buffer(currentRequestPolygon, preFetchDistance , 'miles');
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: changed

File: twittermap/public/javascripts/cache/services.js [hunk 28]
```diff
@@ -0,0 +1,321 @@
...
+           features[id].properties["centerLat"] = (features[id].maxY + features[id].minY) / 2;
+           treeID = box[0]+"/"+box[1]+"/"+box[2]+"/"+box[3];
+           if( insertedTreeIDs.has(treeID) == false){
+                   nodes.push(features[id]);
+                   insertedTreeIDs.add(treeID);
+
+           }
+
+       }
+       //Cachesize already present cachesize,We need to add nodes.length polygons,CacheThreshold is the max size of Cache
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Changed Now will be clear

File: twittermap/public/javascripts/cache/services.js [hunk 29]
```diff
@@ -0,0 +1,321 @@
...
+       var C_maxX = cache_bbox[2];
+       var C_maxY = cache_bbox[3];
+
+       var request_bbox = turf.bbox(currentRequest);
+       var R_minX = request_bbox[0];
+       var R_minY = request_bbox[1];
+       var R_maxX = request_bbox[2];
+       var R_maxY = request_bbox[3];
+
+       var CacheMBR = turf.bboxPolygon(cache_bbox);
```
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Changed

File: twittermap/public/javascripts/cache/services.js (hunk 17 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Changed

File: twittermap/public/javascripts/cache/services.js (hunk 19 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Rephrased

File: twittermap/public/javascripts/cache/services.js (hunk 7 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Added

File: twittermap/public/javascripts/cache/services.js (hunk 11 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: rephrased

File: twittermap/public/javascripts/cache/services.js (hunk 13 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: changed

File: twittermap/public/javascripts/cache/services.js (hunk 14 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: changed

File: twittermap/public/javascripts/cache/services.js (hunk 15 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: changed

File: twittermap/public/javascripts/cache/services.js (hunk 16 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Added Comment

File: twittermap/public/javascripts/cache/services.js (hunk 20 above)
Comment (vidhya567) [inline twittermap/public/javascripts/cache/services.js]: Oh Ok You want to change the comment, I did
